integrate now allows fps

leosartaj opened the pull request:
<!-- Please give this pull request a descriptive title. Pull requests with descriptive titles are more likely to receive reviews. Describe what you changed! A title that only references an issue number is not descriptive. -->

<!-- If this pull request fixes an issue please indicate which issue by typing "Fixes #NNNN" below. -->
Fixes #9885

leosartaj committed:
integrate now allows fps

leosartaj committed:
update doc examples

leosartaj committed:
move logic to doit method

leosartaj committed:
switch of piecewise_fold as default

leosartaj committed:
remove print statement

leosartaj committed:
switch off piecewise_folding as default in ExprWithLimits

leosartaj committed:
remove extra lines

jksuom commented on sympy/integrals/integrals.py:
```diff
@@ -372,6 +372,7 @@ def doit(self, **hints):
         sympy.integrals.rationaltools.ratint
         as_sum : Approximate the integral using a sum
         """
+        print(self.limits)
```
Is this for debugging?

leosartaj replied:
Ahh, I will get this removed.

leosartaj committed:
remove print statements

leosartaj committed:
import integrate in docstrings

smichr commented:
@asmeurer , see my edits in my leo branch for the way I would hackishly handle this.

asmeurer commented:
Not entirely clear what all is being done there but if piecewise specific stuff is in Integral instead of the ExprWithLimits superclass, that's an improvement.

leosartaj committed:
ignore FormalPowerSeries in piecewise_fold

leosartaj commented:
@asmeurer @smichr I found a way to get this working. I went with piecewise_fold just ignoring `FormalPowerSeries` objects. Can some one take a look. Might be there is a better way.

smichr commented on sympy/integrals/integrals.py:
```diff
@@ -401,6 +402,12 @@ def doit(self, **hints):
         if isinstance(function, MatrixBase):
             return function.applyfunc(lambda f: self.func(f, self.limits).doit(**hints))
 
+        if isinstance(function, FormalPowerSeries):
```
I think the integrate method of FormalPowerSeries should be renamed `_eval_integral` and it should accept all the hints that integration accepts. Putting this here is hackishly working around the fact that the `_eval_integral` method has not been defined.

I don't fully understand why the folded version of the fps is failing to integrate unless it's because integration doesn't yet know what to do with the mod expression (in the failing test). Since it seems like folding affects integration it seems better to handle the decision to fold or not to fold in the integration routine as I did for this issue in #12587.

leosartaj replied:
> I think the integrate method of FormalPowerSeries should be renamed _eval_integral and it should accept all the hints that integration accepts. Putting this here is hackishly working around the fact that the _eval_integral method has not been defined.

I agree it should be made cleaner.

> I don't fully understand why the folded version of the fps is failing to integrate unless it's because integration doesn't yet know what to do with the mod expression (in the failing test). 

It's because there are no algorithms written for identifying and integrating `FormalPowerSeries`. Mostly because `FormalPowerSeries` has a different representation. It saves the formulas for the series expansion as `Piecewise` expressions. So integration by default doesn't know what to do with `FormalPowerSeries` object.

> Since it seems like folding affects integration it seems better to handle the decision to fold or not to fold in the integration routine as I did for this issue in #12587.

Okay. I will take a look and try to come up with a better solution.

smichr replied:
One thought that comes to me that I haven't evaluated fully is whether folding for the purposes of integration should only fold Piecewise args that contain the integration variable. This case applies to the integration at hand because the fps has k-dependent Piecewise whereas the integration is being done over x. This could be handled by adding a `x=None` arg to `piecewise_fold`. The default would do what is currently done. If x were not None then it would first mask off all Piecewise, `pw` for which `pw.free_symbols & x` was true. Not sure if x should only be a symbol or not. What would happen with multiple integrations in different variables? Maybe x could be None, a symbol or an iterable and internally it would be converted to a set of free symbols, e.g. x.free_symbols or `Tuple(*x).free_symbols` when x is given as an iterable.

leosartaj replied:
Your latest commit takes away the need to check for `FormalPowerSeries` input to `piecewise_fold`. I will try to take this out and put it into a `_eval_integral` type method.

smichr committed:
extract common Integral.__new__ code

handle fps by the general method of not
folding non-symbol-of-integration-containing
Piecewise args.

leosartaj committed:
Merge pull request #1 from smichr/leo

extract common Integral.__new__ code

smichr commented:
Let's go ahead and make the introduction of `_eval_integral` a separate PR. I'll commit this as is for now.

leosartaj commented:
Okay. Thanks!

On Aug 28, 2017 5:11 PM, "Christopher Smith" <notifications@github.com>
wrote:

> Merged #12976 <https://github.com/sympy/sympy/pull/12976>.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/sympy/sympy/pull/12976#event-1223916609>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AF7g-CpJJgmJl9yj3IabHCXMILO5A4FNks5scti9gaJpZM4OZEHC>
> .
>
